sensor: bq274xx: Add BQ27421 driver

parthitce opened the pull request:
Add support for TI BQ27421 fuel gauge sensor

Signed-off-by: Parthiban Nallathambi <parthiban@linumiz.com>
Signed-off-by: NavinSankar Velliangiri <navin@linumiz.com>

Based on https://github.com/zephyrproject-rtos/zephyr/issues/2542#issuecomment-400299162

MaureenHelm commented on drivers/sensor/CMakeLists.txt:
```diff
@@ -15,6 +15,7 @@ add_subdirectory_ifdef(CONFIG_BME680		bme680)
 add_subdirectory_ifdef(CONFIG_BMG160		bmg160)
 add_subdirectory_ifdef(CONFIG_BMI160		bmi160)
 add_subdirectory_ifdef(CONFIG_BMM150		bmm150)
+add_subdirectory_ifdef(CONFIG_BQ274XX           bq274xx)
```
Please replace the spaces with tab(s) to be consistent with the rest of the file

parthitce replied:
Done thanks.

MaureenHelm commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -0,0 +1,451 @@
+
+
+static int bq274xx_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
```
Need to parse `chan` and only read the channel(s) requested

parthitce replied:
Done thanks.

MaureenHelm commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -0,0 +1,451 @@
+ *
+ * @return 0 for success
+ */
+int bq274xx_gauge_init(struct device *dev)
```
make this function `static`

parthitce replied:
Done thanks.

MaureenHelm commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -0,0 +1,451 @@
+	}
+
+	for (u8_t i = 0; i < 32; i++)
+		block[i] = 0;
```
Wrap in curly braces

parthitce replied:
Done thanks

MaureenHelm commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -0,0 +1,451 @@
+{
+	struct bq274xx_data *bq274xx = dev->driver_data;
+	int status = 0;
+	u8_t tmp_checksum = 0, checksumOld = 0, checksumNew = 0;
```
please no camelcase. see https://docs.zephyrproject.org/latest/contribute/index.html#coding-style

parthitce replied:
Done thanks.

MaureenHelm commented on samples/sensor/bq274xx/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
```
please add the sample in a separate commit in the same PR

parthitce replied:
Done thanks.

MaureenHelm commented on samples/sensor/bq274xx/prj.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_BQ274XX=y
+
+CONFIG_LOG=y
+CONFIG_LOG_IMMEDIATE=y
```
is this really needed?

parthitce replied:
Not needed as it is not consumed. Removed, thanks

MaureenHelm commented on samples/sensor/bq274xx/src/main.c:
```diff
@@ -0,0 +1,186 @@
+		} else {
+			printk("Standby Current: %d.%06dA\n", current_standby.val1,
+					current_standby.val2);
+		}
```
extract this into a function so you don't have to repeat it multiple times

parthitce replied:
Abstracted out. Thanks

MaureenHelm reviewed (changes requested):
Please fix the kconfig and shippable failures

gmarull commented on drivers/sensor/bq274xx/Kconfig:
```diff
@@ -0,0 +1,32 @@
+
+config BQ274XX_TERMINATE_VOLTAGE
+	int "Battery Terminate Voltage in mV"
+	default 3000
```
I think these settings should be part of a DeviceTree binding

parthitce replied:
@gmarull Thanks. These default values are to configure the sensor during initialization. Should the init values/defaults go to device tree?

gmarull replied:
According to https://docs.zephyrproject.org/latest/guides/dts/dt-vs-kconfig.html#dt-vs-kconfig I think that these values should go into DT.

parthitce replied:
Resolved, thanks.

parthitce committed:
sensor: bq274xx: Add BQ27421 driver

Add support for TI BQ27421 fuel gauge sensor

Signed-off-by: Parthiban Nallathambi <parthiban@linumiz.com>
Signed-off-by: NavinSankar Velliangiri <navin@linumiz.com>

parthitce committed:
samples: sensor: bq274xx: Add BQ27421 sample application

Add bq274xx sample with reading and showing all
possible values from the sensor.

Signed-off-by: Parthiban Nallathambi <parthiban@linumiz.com>
Signed-off-by: NavinSankar Velliangiri <navin@linumiz.com>

parthitce commented:
> Please fix the kconfig and shippable failures

Thanks. This PR depends on https://github.com/zephyrproject-rtos/zephyr/pull/23538 for platform_whitelist and adding overlay in the sample.

MaureenHelm commented on drivers/sensor/bq274xx/bq274xx.h:
```diff
@@ -0,0 +1,209 @@
+
+
+
+int bq274xx_command_reg_read(struct bq274xx_data *bq274xx, u8_t reg_addr,
```
why do you have function implementations in the header file? please move them to .c

parthitce replied:
Moved to .c, thanks.

MaureenHelm reviewed (changes requested):
Please add this sensor to `tests/drivers/build_all/sensors_a_h.conf`

parthitce commented:
> Please add this sensor to `tests/drivers/build_all/sensors_a_h.conf`

Added to test, thanks.

gmarull commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -0,0 +1,603 @@
+#if defined(DT_INST_0_TI_BQ274XX_DESIGN_VOLTAGE)
+#define BQ274XX_DESIGN_VOLTAGE		DT_INST_0_TI_BQ274XX_DESIGN_VOLTAGE
+#else
+#define BQ274XX_DESIGN_VOLTAGE		3700
```
You can set a `default` in the DT binding instead. See https://docs.zephyrproject.org/latest/guides/dts/bindings.html#bindings-file-syntax

parthitce replied:
Adapter, thanks for pointing it.

gmarull commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -0,0 +1,603 @@
+		return -EIO;
+	}
+
+	k_sleep(5);
```
Constants should be defined and documented if applicable (e.g. is `5` a recommended value by vendor?)

parthitce replied:
Yes, delay in needed to wait for the data to be stable. Moved to macro. Thanks.

gmarull commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -0,0 +1,603 @@
+	/* Seal the gauge */
+	status = bq274xx_control_reg_write(bq274xx, BQ274XX_CONTROL_SEALED);
+	if (status < 0) {
+		LOG_ERR("Failed to seal the gauge\n");
```
`\n` not needed

parthitce replied:
Done, thanks

gmarull commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -0,0 +1,584 @@
+		return -EIO;
+	}
+
+	designcap_msb = DT_INST_0_TI_BQ274XX_DESIGN_CAPACITY >> 8;
```
In case you plan to support more than a single instance, the `cfg_info` from `DEVICE_AND_API_INIT` field can be used to store constant configuration. This way driver code does not depend on the instance number.

parthitce replied:
@gmarull I understand moving the static DT properties to cfg_info makes sense. But I don't understand "support more than a single instance" for I2C device?

gmarull replied:
Current driver code can only handle a single sensor instance. In order to account for any number of instances it is better to decouple driver code from DT constants. One way to do that is to store DT constants into the `static const` configuration structure and then provide a custom initializer, e.g.

```c
#define MYSENSOR_INIT(index)                                        \
                                                                    \
    static const struct mysensor_config mysensor_config_##index = { \
	    .prop_name = DT_INST_PROP(index, prop_name)                 \
    };                                                              \
                                                                    \
    static struct mysensor_data mysensor_data_##index = {           \
	    .other_prop_name = DT_INST_PROP(index, other_prop_name)     \
    };                                                              \
                                                                    \
    DEVICE_AND_API_INIT(mysensor_##index, DT_INST_LABEL(index),     \
		    &mysensor_init,                                         \
		    &mysensor_data_##index, &mysensor_cfg_##index,          \
		    POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY,               \
		    &mysensor_api);                                         \
									                                 

DT_INST_FOREACH(MYSENSOR_INIT)
```

parthitce replied:
@gmarull Thanks for the details. Fixed.

gmarull commented on samples/sensor/bq274xx/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Replace with ` find_package(Zephyr HINTS $ENV{ZEPHYR_BASE}) ` (recently introduced)

parthitce replied:
Done thanks.

gmarull commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -0,0 +1,584 @@
+	     terminatevolt_msb, terminatevolt_lsb, taperrate_msb, taperrate_lsb;
+	u8_t block[32];
+
+	designenergy_mwh = (u16_t) 3.7 * DT_INST_0_TI_BQ274XX_DESIGN_CAPACITY;
```
New DT macros should be used, e.g. in this case `DT_INST_PROP(0, design_capacity)`

parthitce replied:
Replaced, thanks.

gmarull commented:
Latest changes look good @parthitce. I think there's some problem with tests on CI now.

parthitce commented:
> Latest changes look good @parthitce. I think there's some problem with tests on CI now.

@gmarull Thanks. Am still trying to figure out the new usage of app.overlay. Test entries are moved to app.overlay and this sensor is only applicable for nrf9160_innblue21 board (which is yet to be merged https://github.com/zephyrproject-rtos/zephyr/pull/23538). But the compilation failed for frdm board, which I couldn't understand still.

parthitce commented:
> Latest changes look good @parthitce. I think there's some problem with tests on CI now.

@gmarull @MaureenHelm CI test failure fixed. Please let me know further comments if any. Thanks.

gmarull reviewed (approved):
Thanks for the changes, approved. NOTE: some checkpatch warnings may be easily fixed using `clang-format` on your code.

MaureenHelm reviewed (changes requested):
Please fix the checkpatch warnings

parthitce commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -0,0 +1,598 @@
+	u8_t i2c_data[2];
+	int status;
+
+	status = i2c_burst_read(bq274xx->i2c, BQ274XX_I2C_ADDRESS,
```
Thanks, fixed.

parthitce commented on samples/sensor/bq274xx/src/main.c:
```diff
@@ -0,0 +1,140 @@
+	printk("device is %p, name is %s\n", dev, dev->config->name);
+
+	do_main(dev);
+}
```
Added, thanks.

parthitce commented on samples/sensor/bq274xx/src/main.c:
```diff
@@ -0,0 +1,140 @@
+{
+	struct device *dev;
+
+	dev = device_get_binding(DT_INST_0_TI_BQ274XX_LABEL);
```
Fixed, thanks

parthitce commented:
> Thanks for the changes, approved. NOTE: some checkpatch warnings may be easily fixed using `clang-format` on your code.

@gmarull Thanks the hint, that helped a lot.

parthitce commented:
> Please fix the checkpatch warnings

@MaureenHelm Fixed thanks.

olsky commented:
Great! Thank you @MaureenHelm :-)
